File: bindings/vbnet/src/EveraniumItb.VisualBasic/Status.vb
' Status codes mirrored from the libitb C ABI
' (cmd/cshared/internal/capi/errors.go; numeric values are stable
' across releases), the exception surface for structured
' Try ... Catch handling, and the internal relay translating the C#
' layer's exception into this binding's own type.

''' <summary>Integer status code carried by every
''' <see cref="ItbException"/>.</summary>
Public Enum Status
    Ok = 0
    BadHash = 1
    BadKeyBits = 2
    BadHandle = 3
    BadInput = 4
    BufferTooSmall = 5
    EncryptFailed = 6
    DecryptFailed = 7
    SeedWidthMix = 8
    BadMac = 9
    MacFailure = 10
    BlobMalformedRecipe = 11
    RecipePrimitiveUnknown = 12
    UnknownProfile = 13
    Reserved14 = 14
    Reserved15 = 15
    Reserved16 = 16
    Reserved17 = 17
    BlobModeMismatch = 19
    BlobMalformed = 20
    BlobVersionTooNew = 21
    BlobTooManyOpts = 22
    StreamTruncated = 23
    StreamAfterFinal = 24
    TripleClosed = 25
    ProfileExists = 26
    Internal = 99
End Enum

''' <summary>
''' Raised when libitb reports a non-OK status. <see cref="Status"/>
''' carries the structural code; <see cref="Exception.Message"/>
''' relays the C# layer's formatted diagnostic (which appends the
''' <c>ITB_LastError</c> text captured immediately after the failing
''' call — process-global last-write-wins, so under concurrent use
''' the message may belong to a different call; the status code is
''' always attributable). The originating C# exception rides on
''' <see cref="Exception.InnerException"/>.
''' </summary>
Public NotInheritable Class ItbException
    Inherits Exception

    Private ReadOnly _status As Status

    ''' <summary>The libitb status code for the failing call.</summary>
    Public ReadOnly Property Status As Status
        Get
            Return _status
        End Get
    End Property

    Friend Sub New(status As Status, inner As Global.Itb.ItbException)
        MyBase.New(inner.Message, inner)
        _status = status
    End Sub
End Class

''' <summary>Translation + guard helpers shared by every delegating
''' member of the binding: a C# <c>Itb.ItbException</c> escaping the
''' inner layer is re-raised as this binding's
''' <see cref="ItbException"/> with the numeric status preserved.
''' </summary>
Friend Module ErrorRelay

    Friend Function Translate(ex As Global.Itb.ItbException) As ItbException
        ' The numeric values are identical on both sides (both mirror
        ' the C ABI); an unlisted code passes through as its raw value.
        Return New ItbException(CType(CInt(ex.Status), Status), ex)
    End Function

    Friend Function Guarded(Of T)(fn As Func(Of T)) As T
        Try
            Return fn()
        Catch ex As Global.Itb.ItbException
            Throw Translate(ex)
        End Try
    End Function

    Friend Sub Guarded(fn As Action)
        Try
            fn()
        Catch ex As Global.Itb.ItbException
            Throw Translate(ex)
        End Try
    End Sub
End Module

File: bindings/python/itb3/status.py
"""Status codes mirrored from the libitb3 C ABI
(``cmd/cshared/internal/capi/errors.go``). Numeric values are stable
across releases.
"""

from __future__ import annotations

import enum


class Status(enum.IntEnum):
    """Integer status code returned by every libitb3 entry point."""

    OK = 0
    BAD_HASH = 1
    BAD_KEY_BITS = 2
    BAD_HANDLE = 3
    BAD_INPUT = 4
    BUFFER_TOO_SMALL = 5
    ENCRYPT_FAILED = 6
    DECRYPT_FAILED = 7
    SEED_WIDTH_MIX = 8
    BAD_MAC = 9
    MAC_FAILURE = 10
    BLOB_MALFORMED_RECIPE = 11
    RECIPE_PRIMITIVE_UNKNOWN = 12
    UNKNOWN_PROFILE = 13
    RESERVED_14 = 14
    RESERVED_15 = 15
    RESERVED_16 = 16
    RESERVED_17 = 17
    BLOB_MODE_MISMATCH = 19
    BLOB_MALFORMED = 20
    BLOB_VERSION_TOO_NEW = 21
    BLOB_TOO_MANY_OPTS = 22
    STREAM_TRUNCATED = 23
    STREAM_AFTER_FINAL = 24
    TRIPLE_CLOSED = 25
    PROFILE_EXISTS = 26
    INTERNAL = 99


def status_from(code: int) -> Status:
    """Maps a raw return code onto :class:`Status`; unknown codes
    collapse to :attr:`Status.INTERNAL`."""
    try:
        return Status(code)
    except ValueError:
        return Status.INTERNAL
